### PR TITLE
Show & Tell: Add mod comments

### LINKED ISSUE #716: [Idea]: Mod-only comments for Show & Tell posts
### Describe your idea / enhancement

Allow mods to add mod-only comments to posts (e.g. "held back review cause x")

### How would this idea / enhancement be beneficial?

Help collaborating on reviews

### LINKED ISSUE #717: [Idea]: Public mod comments on Show & Tell posts
### Describe your idea / enhancement

Allow mods to add public comments to show and tell posts ("[removed x]"), not counting towards the char limit.

### How would this idea / enhancement be beneficial?

Allow transparent edits where changes are made public.

### PR DESCRIPTION
**Pull request by ACA30**
## Describe your changes

Resolves #716, resolves #717. Adds support for both public & private mod comments on Show & Tell posts.

image
image

## Todo
- [x] Add support for mod comments
- [x] Display mod comments in admin area
- [x] Display mod comments in frontend
- [x] Make it pretty

### CODE REVIEW

**Comment by ACA30** [comment]
image

**Comment by pjeweb** [comment]
wow. this looks great! i'll set up a deployment with the database schema changes as soon as i can and then review it in detail

**Comment by ACA30** [comment]
@pjeweb saw you reviewing some other PRs - just wanted to check in and see if you had the chance to review this one!

File: apps/website/src/components/show-and-tell/ShowAndTellEntry.tsx [hunk 1]
```diff
@@ -191,6 +195,14 @@ const Content = ({ entry, isPresentationView }: ShowAndTellEntryProps) => {
     }
   }, [entry.text, entry.id]);
 
+  const { data: modComments, error } = trpc.showAndTell.getModComments.useQuery(
+    entry.id,
+  );
+
+  if (error) {
+    console.error("Failed to fetch mod comments:", error);
+  }
```
Comment (pjeweb) [inline apps/website/src/components/show-and-tell/ShowAndTellEntry.tsx:204]: I assume you made this a separate query 'cause fetching them with the posts was messy with types or something? If you can make it work in the main post data a bunch of the following comments will be obsolete.

File: apps/website/src/server/db/show-and-tell.ts [hunk 2]
```diff
@@ -220,6 +220,11 @@ export async function getPostById(
     include: {
       ...withAttachments.include,
       user: true,
+      modComments: {
+        include: {
+          user: true,
+        },
+      },
```
Comment (pjeweb) [inline apps/website/src/server/db/show-and-tell.ts:227]: This makes it easy to leak the internal mod comments and user data for the mods (including their email). So ideally we should have separate functions for the admin features and I'd probably want this to only explicitly select the username and not the other user fields.

File: apps/website/src/server/db/show-and-tell.ts [hunk 3]
```diff
@@ -498,3 +503,46 @@ export async function getPostsToShow() {
...
   return postsToShow;
 }
+
+// Fetch all mod comments for a specific entry
+export async function getModCommentsForEntry(entryId: string) {
+  return await prisma.showAndTellModComment.findMany({
+    where: { entryId },
+    include: { user: true },
+  });
+}
```
Comment (pjeweb) [inline apps/website/src/server/db/show-and-tell.ts:513]: Dito used for both admin and public data, fetches *all* mod comments (including internal ones) and selects all fields from the mod user (incl. email)! Public data should not contain any mod user info, admin data only the username.

File: apps/website/src/server/trpc/router/admin/show-and-tell.ts [hunk 4]
```diff
@@ -29,6 +33,54 @@ export const adminShowAndTellRouter = router({
...
+    .input(z.string().cuid())
+    .query(async ({ input }) => {
+      const comments = await getModCommentsForEntry(input);
+      if (!comments) {
+        throw new TRPCError({
+          code: "NOT_FOUND",
+          message: "Comments not found",
+        });
+      }
+      return comments;
```
Comment (pjeweb) [inline apps/website/src/server/trpc/router/admin/show-and-tell.ts:46]: Is the explicit 404 needed or can this just return the empty result?
```suggestion
      return getModCommentsForEntry(input);
```

File: apps/website/src/server/trpc/router/admin/show-and-tell.ts [hunk 5]
```diff
@@ -29,6 +33,54 @@ export const adminShowAndTellRouter = router({
...
+      }),
+    )
+    .mutation(async ({ input }) => {
+      return await addModComment(
+        input.entryId,
+        input.modId,
+        input.comment,
+        input.isInternal,
+      );
+    }),
```
Comment (pjeweb) [inline apps/website/src/server/trpc/router/admin/show-and-tell.ts:65]: Nit pick but we can spare one asyc/await promise resolving by just returning the promise directly here.

```suggestion
    .mutation(({ input }) => addModComment(
      input.entryId,
      input.modId,
      input.comment,
      input.isInternal,
    )),
```

File: apps/website/src/server/trpc/router/show-and-tell.ts [hunk 6]
```diff
@@ -147,4 +148,17 @@ export const showAndTellRouter = router({
...
+    .input(z.string().cuid())
+    .query(async ({ input }) => {
+      const comments = await getModCommentsForEntry(input);
+      if (!comments) {
+        throw new TRPCError({
+          code: "NOT_FOUND",
+          message: "Comments not found",
+        });
+      }
+      return comments.filter((comment) => !comment.isInternal);
```
Comment (pjeweb) [inline apps/website/src/server/trpc/router/show-and-tell.ts:162]: Again make sure we only pass public data here (= no mod user info).

**Review by pjeweb — CHANGES_REQUESTED** [review]
Looks great already. I only tested locally right now. Let me know if we need it on a public deployment for reviewing!

File: apps/website/src/pages/admin/show-and-tell/review/[entryId]/index.tsx [hunk 7]
```diff
@@ -154,6 +212,75 @@ const AdminReviewShowAndTellPage: NextPage<
...
+                  value={newComment}
+                  onChange={(e) => setNewComment(e.target.value)}
+                  required
+                />
+                <button
+                  type="submit"
+                  className="flex items-center justify-center rounded-r p-2 text-white"
+                >
+                  <IconPaperAirplane className="h-6 w-6" />
+                </button>
```
Comment (pjeweb) [inline apps/website/src/pages/admin/show-and-tell/review/[entryId]/index.tsx:278]: Can this get a label? We usually give buttons a visual label even tho I like the icon. Also good for accessibility ;)

**Comment by ACA30** [comment]
Awesome, tysm! Will make all these updates :)

**Comment by MattIPv4** [comment]
No progress, closing.

**Pull request closed without merging**